core/lib/vm_executor/src/oneshot/mod.rs

```diff
@@ -182,25 +254,89 @@ where
     }
 }
 
+#[derive(Debug)]
+enum Vm<S: ReadStorage> {
+    Legacy(LegacyVmInstance<S, HistoryDisabled>),
+    Fast(FastVmInstance<S, ()>),
+}
+
+impl<S: ReadStorage> Vm<S> {
+    fn inspect_transaction_with_bytecode_compression(
+        &mut self,
+        missed_storage_invocation_limit: usize,
+        params: OneshotTracingParams,
+        tx: Transaction,
+        with_compression: bool,
+    ) -> OneshotTransactionExecutionResult {
+        let mut calls_result = Arc::<OnceCell<_>>::default();
+        let (compression_result, tx_result) = match self {
+            Self::Legacy(vm) => {
+                let mut tracers = Self::create_legacy_tracers(
+                    missed_storage_invocation_limit,
+                    params.trace_calls.then(|| calls_result.clone()),
+                );
+                vm.inspect_transaction_with_bytecode_compression(&mut tracers, tx, with_compression)
+            }
+            Self::Fast(vm) => {
+                assert!(
+                    !params.trace_calls,
+                    "Call tracing is not supported by fast VM yet"
+                );
+                let legacy_tracers = Self::create_legacy_tracers::<HistoryEnabled>(
+                    missed_storage_invocation_limit,
+                    None,
+                );
+                let mut full_tracer = (legacy_tracers.into(), ());
+                vm.inspect_transaction_with_bytecode_compression(
+                    &mut full_tracer,
+                    tx,
+                    with_compression,
+                )
+            }
+        };
+
+        OneshotTransactionExecutionResult {
+            tx_result: Box::new(tx_result),
+            compression_result: compression_result.map(drop),
+            call_traces: Arc::make_mut(&mut calls_result).take().unwrap_or_default(),
+        }
+    }
+
+    fn create_legacy_tracers<H: HistoryMode>(
+        missed_storage_invocation_limit: usize,
+        calls_result: Option<Arc<OnceCell<Vec<Call>>>>,
+    ) -> TracerDispatcher<StorageView<S>, H> {
+        let mut tracers = vec![];
+        if let Some(calls_result) = calls_result {
+            tracers.push(CallTracer::new(calls_result).into_tracer_pointer());
+        }
+        tracers
+            .push(StorageInvocations::new(missed_storage_invocation_limit).into_tracer_pointer());
+        tracers.into()
+    }
+}
+
 #[derive(Debug)]
 struct VmSandbox<S: ReadStorage> {
-    vm: Box<LegacyVmInstance<S, HistoryDisabled>>,
-    storage_view: StoragePtr<StorageView<S>>,
+    vm: Vm<StorageWithOverrides<S>>,
```

Comment:
I think the VM can built when something is executed, so it doesn't need to live in this struct.